Fix #6888: Run tests split by classes

ankita240796 opened the pull request:
## Explanation
Fixes #6888: Splits running backend tests by test classes instead of test files. This will help to speed up the backend tests hopefully (Currently the tests take 3 hours to run, so once backend tests finish running on this PR, we can check if splitting by classes is doing a help here)

## Checklist
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python scripts/pre_commit_linter.py` and `bash scripts/run_frontend_tests.sh`.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR has an appropriate "CHANGELOG: ..." label (If you are unsure of which label to add, ask the reviewers for guidance).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [x] The PR addresses the points mentioned in the codeowner checks for the files/folders changed. (See the [codeowner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).)
- [x] The PR is **assigned** to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer and don't tick this checkbox.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue. Do not only request the review but also add the reviewer as an assignee.

ankita240796 committed:
Run tests split by classes

ankita240796 committed:
Fix Test/Tests issue

ankita240796 commented:
Yes!!! I think this fixes the issue since it took only 30 minutes to complete the backend tests!

apb7 commented:
Hi @ankita240796, this looks good but is there a way to actually reduce the running time of the test causing the issue? Backend tests usually used to take around 19 minutes on CircleCI. Thanks!

ankita240796 commented:
Hi @apb7, I don’t think we can reduce the time further. Since the other fix we can have is to split the test file into multiple test file one per storage model but that will basically be the same as running the test split by classes. The prod jobs test will take some increase in time since the jobs are being run for each test function.

Also, I am not sure why the coverage is being shown as decreased here since I haven’t made changes to the file which are being reported. Will merging develop fix that issue?

Thanks!

apb7 commented:
> Hi @apb7, I don’t think we can reduce the time further. Since the other fix we can have is to split the test file into multiple test file one per storage model but that will basically be the same as running the test split by classes. The prod jobs test will take some increase in time since the jobs are being run for each test function.

Okay, thanks for the explanation. This looks good then.  

> Also, I am not sure why the coverage is being shown as decreased here since I haven’t made changes to the file which are being reported. Will merging develop fix that issue?

Yes, that should help. /cc @lilithxxx

Thanks!

ankita240796 committed:
Fix naming

ankita240796 commented:
Hi @apb7, I just checked the branch is already updated. It was naming issue in the test files. I have fixed that now, PTAL. Thanks!

apb7 commented on core/domain/exp_jobs_one_off_test.py:
```diff
@@ -2178,7 +2178,7 @@ def test_no_action_is_performed_for_deleted_exploration(self):
 
 
-class TestDeleteStateIdMappingModelsOneOffJob(test_utils.GenericTestBase):
+class DeleteStateIdMappingModelsOneOffJobTests(test_utils.GenericTestBase):
```
I think there are other files as well which follow this convention. Can you do a grep for `class Test`?

ankita240796 replied:
Yes, I checked that already but they are mock classes or base classes and are not being used in running the test.

apb7 reviewed (approved):
LGTM, thanks @ankita240796!

seanlip commented on scripts/backend_tests.py:
```diff
@@ -211,7 +211,24 @@ def _get_all_test_targets(test_path=None, include_load_tests=True):
     def _convert_to_test_target(path):
         """Remove the .py suffix and replace all slashes with periods."""
-        return os.path.relpath(path, os.getcwd())[:-3].replace('/', '.')
+        classes = []
```
class_names

ankita240796 replied:
Done.

seanlip commented on scripts/backend_tests.py:
```diff
@@ -211,7 +211,24 @@ def _get_all_test_targets(test_path=None, include_load_tests=True):
     """
     def _convert_to_test_target(path):
         """Remove the .py suffix and replace all slashes with periods."""
```
Please update the docstring with the correct args, returns, etc. 

Please also change the names of the function / args / variables to reflect what this method is now doing.

ankita240796 replied:
Done.

ankita240796 replied:
Done.

seanlip commented on scripts/backend_tests.py:
```diff
@@ -211,7 +211,24 @@ def _get_all_test_targets(test_path=None, include_load_tests=True):
+            for line in lines:
+                if line.startswith('class'):
+                    start_index = line.find(' ') + 1
+                    end_index = line.find('Tests(')
```
I think this is fragile. It just takes someone to not name their class properly -- and then the test won't run, but no one will notice it.

Instead, could you do a check based on what the class is a subclass of? You might be able to use stuff like inspect.getmro (grep for that in the codebase) to do this more semantically, but I think the correct definition of test classes is "is a subclass of the base test class", not "is named XXX".

ankita240796 replied:
I actually tried using importlib to import module from the test_target_path but I get errors there as follows:

```
Checking if node.js is installed in /home/ankita/Desktop/opensrc/oppia_tools
Checking whether Google App Engine is installed in /home/ankita/Desktop/opensrc/oppia_tools/google_appengine_1.9.67/google_appengine
Checking whether google-cloud-sdk is installed in /home/ankita/Desktop/opensrc/oppia_tools/google-cloud-sdk-251.0.0/google-cloud-sdk
Traceback (most recent call last):
  File "scripts/backend_tests.py", line 395, in <module>
    main()
  File "scripts/backend_tests.py", line 290, in main
    include_load_tests=include_load_tests)
  File "scripts/backend_tests.py", line 246, in _get_all_test_targets
    _convert_to_test_target(os.path.join(base_path, root)))
  File "scripts/backend_tests.py", line 219, in _convert_to_test_target
    python_module = importlib.import_module(test_target_path)
  File "/usr/lib/python2.7/importlib/__init__.py", line 37, in import_module
    __import__(name)
  File "/home/ankita/Desktop/opensrc/oppia/utils_test.py", line 23, in <module>
    from core.tests import test_utils
  File "/home/ankita/Desktop/opensrc/oppia/core/tests/test_utils.py", line 30, in <module>
    from core.domain import collection_services
  File "/home/ankita/Desktop/opensrc/oppia/core/domain/collection_services.py", line 33, in <module>
    from core.domain import exp_services
  File "/home/ankita/Desktop/opensrc/oppia/core/domain/exp_services.py", line 38, in <module>
    from core.domain import email_subscription_services
  File "/home/ankita/Desktop/opensrc/oppia/core/domain/email_subscription_services.py", line 19, in <module>
    from core.domain import email_manager
  File "/home/ankita/Desktop/opensrc/oppia/core/domain/email_manager.py", line 22, in <module>
    from core.domain import config_domain
  File "/home/ankita/Desktop/opensrc/oppia/core/domain/config_domain.py", line 22, in <module>
    import schema_utils
  File "/home/ankita/Desktop/opensrc/oppia/schema_utils.py", line 32, in <module>
    from core.domain import html_cleaner  # pylint: disable=relative-import
  File "/home/ankita/Desktop/opensrc/oppia/core/domain/html_cleaner.py", line 24, in <module>
    import bleach
ImportError: No module named bleach
```

Is there some path issues and do I need to insert path for the modules?

seanlip replied:
Possibly -- see core/tests/gae_suite.py.

But if this becomes too tricky/messy, then perhaps a lexical analysis of the "Abc" in `class XYZ(Abc)` might be fine -- maybe ensure it is from a given set of strings. That's more fragile though, so better to avoid if possible, but it might be a good middle ground if it proves too hard to use getmro().

ankita240796 replied:
I tried adding the paths as in tests/gae_suite.py but I still get errors for modules being imported.

Also, for lexical analysis, the classes should inherit from a fixed class but in many files, we have new base classes and then test classes as a subclass of them. Would it be not better if we could have a test to ensure that all test class names should end in `Test/Tests`?

seanlip replied:
> Also, for lexical analysis, the classes should inherit from a fixed class but in many files, we have new base classes and then test classes as a subclass of them. Would it be not better if we could have a test to ensure that all test class names should end in Test/Tests?

That could also work, but how would you do that? Such a test would still need to identify which classes are "test classes".

And yeah, that is the drawback of lexical analysis, we'd probably have to maintain a list of "allowed subclasses" or something, but that is indeed icky.

brianrodri replied:
Maybe we should use [metaclasses](https://avacariu.me/articles/2016/using-a-metaclass-to-keep-track-of-subclasses) to keep track of subclasses?

ankita240796 replied:
This is actually fixed now. I added paths similar to that in `core/tests/gae_suite.py`. I missed pylint path earlier since it was not present in `gae_suite.py` but test files like `pylint_extensions_tests` require pylint import as well and so adding that to path fixed the issue!

seanlip reviewed (changes requested):
Thank you, @ankita240796! Left some notes, PTAL.

seanlip commented:
(Also, yay on 3 hours --> 30 minutes!!)

ankita240796 committed:
Fix names and comments

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into fix-backend

ankita240796 committed:
Fix docstring

ankita240796 committed:
Fix path issue

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into fix-backend

ankita240796 commented:
Hi @seanlip @brianrodri, I have addressed all the review comments. I am sorry for the delay here, I was actually travelling yesterday and was not able to fix this due to that. PTAL, Thanks!

ankita240796 committed:
Add coverage path to avoid circleci failure

seanlip reviewed (approved):
Woohoo! Thanks @ankita240796 :) LGTM on the general approach, but I think you'll need to figure out why the backend tests are failing. OK to merge once you've got them to pass. Thanks!

ankita240796 commented:
Hi @seanlip, I am trying to check the coverage issue locally but I cannot reproduce it. Is it due to some cache issue on ci? Thanks!

seanlip commented:
TBH, I don't know, sorry...

/cc @apb7 @oppia/dev-workflow-team -- is there a way to clear the cache on CircleCI?

seanlip commented:
@ankita240796, one thought -- if you can open a separate PR with exactly the same changes as this one,  and that PR passes backend tests on CircleCI, then I think I can merge this one. That might be a backup solution...though we should figure out the cache clearing if possible.

ankita240796 commented:
Hi @YashJipkate, I remember you required clearing the ci cache in one of your PRs. If that's the case, can you please guide me here. Thanks!

ankita240796 commented:
> @ankita240796, one thought -- if you can open a separate PR with exactly the same changes as this one, and comment out the other travis lines except the backend tests, and that PR passes backend tests, then I think I can merge this one. That might be a backup solution...

Sure @seanlip, I will do that now. Thanks!

anubhavsinha98 commented:
Hi @ankita240796, I am not sure maybe look you can look at https://support.circleci.com/hc/en-us/articles/115015426888-Clear-project-cache

apb7 commented:
Hi @ankita240796 and @seanlip, I checked out the complete log. This could be caused by cache issues. Ankita, [here](https://github.com/oppia/oppia/wiki/Instructions-for-Maintainers#how-to-clear-caches-for-cis) are the instructions to clear cache on CircleCI. These can be done in this PR only. Ping me in case you have any doubts. Thanks!

ankita240796 committed:
Try clearing cache

ankita240796 commented:
Hi @anubhavsinha98 @apb7, Thanks for the help. I have commented out the lines to clear the cache and we can wait to see if it works as expected.

ankita240796 commented:
It still fails. So, I don't think this is a cache issue. I am trying to search if I can find a fix for this. Thanks!

ankita240796 committed:
Try fixing coverage issue

ankita240796 committed:
Check path

ankita240796 committed:
Remove extra path

ankita240796 committed:
Remove restore and save cache

ankita240796 committed:
Remove print

ankita240796 committed:
Try fixing ci

ankita240796 committed:
Remove cache comments

seanlip commented on scripts/backend_tests.py:
```diff
@@ -24,15 +24,51 @@
+OPPIA_TOOLS_DIR = os.path.join(CURR_DIR, '..', 'oppia_tools')
+THIRD_PARTY_DIR = os.path.join(CURR_DIR, 'third_party')
+DIRS_TO_ADD_TO_SYS_PATH = [
+    os.path.join(OPPIA_TOOLS_DIR, 'coverage-4.5.3', 'coverage', 'cmdline'),
```
It looks like this directory cannot be found. I wonder if it is actually being downloaded and installed correctly? Does the backend test script run the setup/install-third-party scripts first?

seanlip replied:
Maybe you could try doing a listdir to the console to see what's in the folder -- see whether coverage.cmdline exists at all and maybe cat its contents to the terminal.

ankita240796 committed:
Revert to lexical form

seanlip reviewed (approved):
Thanks @ankita240796! Let's get this in to unblock the build, and file an issue for immediate fixing that  tries to do this more robustly.

seanlip commented:
Merging since this passed all checks at https://github.com/oppia/oppia/pull/7040/commits/ca06f27281b20097c5e23d48f22d9eb73e384b3f
